File: examples/18-errors/inventory-service/main.py
import logging
import os
from concurrent import futures

import grpc
from grpc_status import rpc_status
from google.protobuf import any_pb2, duration_pb2
from google.rpc import status_pb2, error_details_pb2

import inventory_pb2
import inventory_pb2_grpc

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("inventory")

INITIAL_STOCK = int(os.getenv("INITIAL_STOCK", "10"))
FAIL_MODE = os.getenv("FAIL_MODE", "none")

stock = {}


class InventoryServicer(inventory_pb2_grpc.InventoryServiceServicer):
    def ReserveStock(self, request, context):
        if FAIL_MODE == "unavailable":
            retry_info = error_details_pb2.RetryInfo(
                retry_delay=duration_pb2.Duration(seconds=2)
            )
            detail = any_pb2.Any()
            detail.Pack(retry_info)
            rich_status = status_pb2.Status(
                code=grpc.StatusCode.UNAVAILABLE.value[0],
                message="inventory service temporarily unavailable",
                details=[detail],
            )
            context.abort_with_status(rpc_status.to_status(rich_status))
            return inventory_pb2.ReserveResponse()

        if request.sku not in stock:
            stock[request.sku] = INITIAL_STOCK

        if stock[request.sku] < request.quantity:
            context.set_code(grpc.StatusCode.FAILED_PRECONDITION)
            context.set_details(f"insufficient stock for {request.sku}: have {stock[request.sku]}, need {request.quantity}")
            return inventory_pb2.ReserveResponse(confirmed=False, remaining=stock[request.sku])

        stock[request.sku] -= request.quantity
        log.info("reserved sku=%s qty=%d remaining=%d", request.sku, request.quantity, stock[request.sku])
        return inventory_pb2.ReserveResponse(confirmed=True, remaining=stock[request.sku])


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    inventory_pb2_grpc.add_InventoryServiceServicer_to_server(InventoryServicer(), server)
    server.add_insecure_port("[::]:50051")
    server.start()
    log.info("inventory-service started on :50051 (stock=%d, fail_mode=%s)", INITIAL_STOCK, FAIL_MODE)
    server.wait_for_termination()


if __name__ == "__main__":
    serve()

File: examples/18-errors/cpp/inventory-service/src/main.cpp
#include <grpcpp/grpcpp.h>
#include <inventory.grpc.pb.h>
#include <spdlog/spdlog.h>
#include <unordered_map>
#include <mutex>
#include <memory>

class InventoryServiceImpl final : public inventory::InventoryService::Service {
private:
    std::unordered_map<std::string, int> stock_;
    std::mutex mutex_;

public:
    InventoryServiceImpl() {
        // Pre-seed limited stock
        stock_["limited"] = 5;
    }

    grpc::Status ReserveStock(grpc::ServerContext* context,
                              const inventory::ReserveRequest* request,
                              inventory::ReserveResponse* response) override {
        std::lock_guard<std::mutex> lock(mutex_);

        const std::string& sku = request->sku();
        int quantity = request->quantity();

        spdlog::info("ReserveStock called: sku={}, quantity={}", sku, quantity);

        // Check if SKU exists in stock map
        auto it = stock_.find(sku);

        if (it != stock_.end()) {
            // SKU exists - check if we have enough
            int available = it->second;

            if (available >= quantity) {
                // Reserve the stock
                it->second -= quantity;
                response->set_confirmed(true);
                response->set_remaining(it->second);
                spdlog::info("Stock reserved: sku={}, remaining={}", sku, it->second);
            } else {
                // Insufficient stock
                response->set_confirmed(false);
                response->set_remaining(available);
                spdlog::warn("Insufficient stock: sku={}, requested={}, available={}",
                           sku, quantity, available);
            }
        } else {
            // Unknown SKU - treat as unlimited stock
            response->set_confirmed(true);
            response->set_remaining(9999);
            spdlog::info("Unknown SKU (unlimited): sku={}", sku);
        }

        return grpc::Status::OK;
    }
};

void RunServer() {
    std::string server_address("0.0.0.0:50051");
    InventoryServiceImpl service;

    grpc::ServerBuilder builder;
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);

    std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
    spdlog::info("Inventory service listening on {}", server_address);
    server->Wait();
}

int main() {
    spdlog::set_level(spdlog::level::info);
    RunServer();
    return 0;
}
